### PR TITLE
Refactor/rest like endpoints

### LINKED ISSUE #143: Refactor endpoints to follow a more "REST-like" approach
Our API currently work fine but there are some endpoint semantics that don't make sense when thinking about the REST standards.
The backend will mostly be an internal tool but it would be great to follow the standards and re-think the API before it starts being heavily used by the frontend (or possibly other services).

You can find a lot of information online about REST best practices and recommendations, but here's one I found useful: https://www.freecodecamp.org/news/rest-api-best-practices-rest-endpoint-design-examples/

I'll also leave some suggestions of recommendations we're not currently following and the respective endpoints.

### Use Nouns Instead of Verbs in Endpoints

This is one of the biggest flaws in our API that was pointed out in one of our meetings. Here are some examples where we violated this (please ensure no more were added in the meantime):
- `POST /accounts/changePassword/{id}` This could probably be changed to something like `POST /accounts/{id}/password`
- `PUT /events/{eventId}/addTeamMember/{accountId}` to `PUT /events/{eventId}/team/{accountId}` or `POST /events/{eventId}/members` with the accountId in the body. Some discussion of a similar problem [here](https://stackoverflow.com/questions/6324547/how-can-i-handle-many-to-many-relationships-in-a-restful-api)
- Same thing as above for removing team members and for projects too. Same for hall of fame (in #139 )
- `PUT /projects/{id}/archive` I'm not sure about this one, I like the endpoint but it violates the "rule". Same for unarchive.

### Redundant words in endpoint path

For example, if we have a POST request, there's no need to include a word like `new` in the path. This is redundant and just complicates things:

- `POST /accounts/new` to `POST /accounts` (and similar ones for creating entities)
- `POST /auth/new` to `POST /auth` maybe

### Use Filtering, Sorting, and Pagination to Retrieve the Data Requested

- Remove `GET /events/category/{category}` and add a query filter to `GET /events` with the same functionality, like `GET /events?category={category}`

### Ambiguous endpoints

It might be a problem to have endpoint paths not clearly explaining what they do:

- `GET /generations` to  `GET /generations/years`

Please give your thoughts about the suggestions and suggest new changes. Be sure to make your own investigations instead of following my suggestions blindly.

### PR DESCRIPTION
**Pull request by Sirze01**
Closes #143 

First of all kudos to @BrunoRosendo for the great groundwork.
This PR refactors some of the endpoints being used and adds support for URL queries in the documentation, as, as of now, some endpoints rely on URL queries to filter the returned contents:

Next is a list of the affected endpoints:
- accounts
    - POST /changePassord/{id} -> POST /{id}/password
    - POST /new -> POST

- auth
    - POST /new -> POST

- events
    - GET /category -> GET ?category=Great Events (Missing documentation for query parameters)
    - POST /new -> POST
    - PUT /{idEvent}/addTeamMember/{idAccount} -> PUT /{idEvent}/team/{idAccount}
    - PUT /{idEvent}/removeTeamMember/{idAccount} -> DELETE /{idEvent}/team/{idAccount}

- generations
    - GET -> GET /years
    - POST /new -> POST

- posts
    - POST new -> POST

- projects
    - POST /new -> POST
    - PUT /{idProject}/addTeamMember/{idAccount} -> PUT /{idProject}/team/{idAccount}
    - PUT /{idProject}/removeTeamMember/{idAccount} -> DELETE /{idProject}/team/{idAccount}
    - PUT /{idEvent}/addHallOffFameMember/{idAccount} -> PUT /{idEvent}/hallOfFameMember/{idAccount}
    - PUT /{idEvent}/removeHallOfFameMember/{idAccount} -> DELETE /{idEvent}/hallOfFameMember/{idAccount}

# Review checklist
-   [x] Properly documents API changes in the corresponding controller test
-   [x] Contains enough appropriate tests
-   [x] Behavior is as expected
-   [x] Clean, well structured code

### CODE REVIEW

**Comment by Sirze01** [comment]
Some endpoints are somewhat rule-breakers.
I thought of some alternatives, if you guys think we should change them, let me know. 

I think having different operations to handle the team and project memberships is somewhat of an anti-pattern in REST terms. Having the operations seems to be way too imperative, almost resembling Remote Procedure Calls (RPC).

- Events
  - Single `PUT /{idEvent}/team` (with a list "team" resource for simplification), the list of members would be rewritten with every update
  - Patching the Event object with the new team members list
  - Just rewriting the whole event object

- Projects
  - Patch to the project object with the new value of `isArchived`
  - `PUT /{id}/` with the modified body in the case of the  archival/unarchival

This is me being pedantic at this point, as I see value in these exceptions, making the endpoints easier to interact with.
Let me know your opinion.

**Comment by BrunoRosendo** [comment]
I haven't reviewed but I want to warn that after #139 this PR needs to address the API on the new endpoints. Let's discuss the team decision in the next meeting

**Comment by LuisDuarte1** [comment]
> Some endpoints are somewhat rule-breakers. I thought of some alternatives, if you guys think we should change them, let me know.
> 
> I think having different operations to handle the team and project memberships is somewhat of an anti-pattern in REST terms. Having the operations seems to be way too imperative, almost resembling Remote Procedure Calls (RPC).
> 
> * Events
>   
>   * Single `PUT /{idEvent}/team` (with a list "team" resource for simplification), the list of members would be rewritten with every update
>   * Patching the Event object with the new team members list
>   * Just rewriting the whole event object
> * Projects
>   
>   * Patch to the project object with the new value of `isArchived`
>   * `PUT /{id}/` with the modified body in the case of the  archival/unarchival
> 
> This is me being pedantic at this point, as I see value in these exceptions, making the endpoints easier to interact with. Let me know your opinion.

I think that it's not pedantic at all and you approach is simpler imo, it simplifies the backend by reducing the number of endpoints, by having the side-effect of sometimes making more api calls (eg.: to add a team member, you have to get the current list and then make the request to rewrite the list of members with the new member. In most situations there's no additional overhead, but in some cases it'll add a request of overhead). I think we need more opinion on this cc: @MRita443

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 1]
```diff
@@ -112,14 +112,94 @@ internal class EventControllerTest @Autowired constructor(
...
+                        |It also allows to filter the events by category, using the query parameter "category".
                         |This is useful for example in the frontend's event page, where events are displayed.
                     """.trimMargin()
                 )
         }
     }
 
+    @NestedTest
+    @DisplayName("GET events?category={category}")
+    inner class GetEventsByCategory {
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]: I think all these tests should be moved into `GetEvents`

**Review by BrunoRosendo — CHANGES_REQUESTED** [review]
Good job with the refactor!

Thoughts/Answers to your comment:
- As for adding members to the teams, I think we could change to POST /events/{id}/team and pass an id in the body, resembling the creation of a new team member. I think this is more RESTful and keeps the API simpler to use
- As for project archival/unarchival, if we want to be very strict about REST norms then we could erase it since we can change it using the regular PUT. The reason I proposed these endpoints at the time was that I could see a clear purpose for them (e.g. clicking an archive button instead of editing a form that updates the project). We can also say that updating and archiving entities are different things (independently of implementation), there seems to be a discussion on this https://stackoverflow.com/questions/16201905/restful-archiving-of-entities-in-webapi
Personaly, I prefer having an exception to REST and keeping the API easily understandable, just like [GitHub](https://docs.github.com/en/free-pro-team@latest/rest/issues/issues?apiVersion=2022-11-28#lock-an-issue) and [Reddit](https://www.reddit.com/dev/api/#POST_api_mod_conversations_:conversation_id_archive) do in some cases.

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt (hunk 1 above)
Comment (BrunoRosendo) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]: This got me thinking that we might want to have more filters on the BE (e.g. archived posts) but I think we can wait to see how FE wants to implement filters

Comment (Sirze01) [reply src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt]: > I think all these tests should be moved into `GetEvents`

Will do!

**Comment by Sirze01** [comment]
> Good job with the refactor!
> 
> Thoughts/Answers to your comment:
> 
>     * As for adding members to the teams, I think we could change to POST /events/{id}/team and pass an id in the body, resembling the creation of a new team member. I think this is more RESTful and keeps the API simpler to use
> 
>     * As for project archival/unarchival, if we want to be very strict about REST norms then we could erase it since we can change it using the regular PUT. The reason I proposed these endpoints at the time was that I could see a clear purpose for them (e.g. clicking an archive button instead of editing a form that updates the project). We can also say that updating and archiving entities are different things (independently of implementation), there seems to be a discussion on this https://stackoverflow.com/questions/16201905/restful-archiving-of-entities-in-webapi
>       Personaly, I prefer having an exception to REST and keeping the API easily understandable, just like [GitHub](https://docs.github.com/en/free-pro-team@latest/rest/issues/issues?apiVersion=2022-11-28#lock-an-issue) and [Reddit](https://www.reddit.com/dev/api/#POST_api_mod_conversations_:conversation_id_archive) do in some cases.

- Regarding team updates
I would go even further and instead of taking only a member ID in the body, take an array of IDs to support batch updates, (Adding/removing more than one in a "commit" action in the frontend)

- Regarding the archive/unarchive
I think we should leave the endpoints as they are now.

**Comment by BrunoRosendo** [comment]
#64 was merged, could you please refactor the endpoints too? It's the exact same thing as adding team members

**Comment by LuisDuarte1** [comment]
@Sirze01 can you look into this please?

**Comment by Sirze01** [comment]
> Closes #143
> 
> First of all kudos to @BrunoRosendo for the great groundwork. This PR refactors some of the endpoints being used and adds support for URL queries in the documentation, as, as of now, some endpoints rely on URL queries to filter the returned contents:
> 
> Next is a list of the affected endpoints:
> 
>     * accounts
>       
>       * POST /changePassord/{id} -> POST /{id}/password
>       * POST /new -> POST
> 
>     * auth
>       
>       * POST /new -> POST
> 
>     * events
>       
>       * GET /category -> GET ?category=Great Events (Missing documentation for query parameters)
>       * POST /new -> POST
>       * PUT /{idEvent}/addTeamMember/{idAccount} -> PUT /{idEvent}/team/{idAccount}
>       * PUT /{idEvent}/removeTeamMember/{idAccount} -> DELETE /{idEvent}/team/{idAccount}
> 
>     * generations
>       
>       * GET -> GET /years
>       * POST /new -> POST
> 
>     * posts
>       
>       * POST new -> POST
> 
>     * projects
>       
>       * POST /new -> POST
>       * PUT /{idProject}/addTeamMember/{idAccount} -> PUT /{idProject}/team/{idAccount}
>       * PUT /{idProject}/removeTeamMember/{idAccount} -> DELETE /{idProject}/team/{idAccount}
>       * PUT /{idEvent}/addHallOffFameMember/{idAccount} -> PUT /{idEvent}/hallOfFameMember/{idAccount}
>       * PUT /{idEvent}/removeHallOfFameMember/{idAccount} -> DELETE /{idEvent}/hallOfFameMember/{idAccount}
> 
> 
> # Review checklist
> 
>     * [x]  Properly documents API changes in the corresponding controller test
> 
>     * [x]  Contains enough appropriate tests
> 
>     * [x]  Behavior is as expected
> 
>     * [x]  Clean, well structured code

Updated the PR description.
The [PR in the frontend seems to be updated still](https://github.com/NIAEFEUP/website-niaefeup-frontend/pull/219).

File: src/main/kotlin/pt/up/fe/ni/website/backend/controller/ProjectController.kt [hunk 2]
```diff
@@ -65,25 +65,25 @@ class ProjectController(private val service: ProjectService) {
...
 
-    @PutMapping("/{idProject}/removeTeamMember/{idAccount}")
+    @DeleteMapping("/{idProject}/team/{idAccount}")
     fun removeTeamMemberById(
         @PathVariable idProject: Long,
         @PathVariable idAccount: Long
     ) = service.removeTeamMemberById(idProject, idAccount)
 
-    @PutMapping("/{idProject}/addHallOfFameMember/{idAccount}")
+    @PutMapping("/{idProject}/hallOfFameMember/{idAccount}")
```
Comment (BrunoRosendo) [inline src/main/kotlin/pt/up/fe/ni/website/backend/controller/ProjectController.kt]:
```suggestion
    @PutMapping("/{idProject}/hallOfFame/{idAccount}")
```
I think this is more similar to "team", wdyt?

**Review by BrunoRosendo — CHANGES_REQUESTED** [review]
really small detail!

File: src/main/kotlin/pt/up/fe/ni/website/backend/controller/ProjectController.kt (hunk 2 above)
Comment (Sirze01) [reply src/main/kotlin/pt/up/fe/ni/website/backend/controller/ProjectController.kt]: Good catch. Already updated it, take a look.

**Review by BrunoRosendo — APPROVED** [review]
Great job!

**Review by LuisDuarte1 — APPROVED** [review]
LGTM, great work :rocket:

**Pull request merged**